src/sql/SqlService.ts

```diff
@@ -359,21 +356,20 @@ export class SqlServiceImpl implements SqlService {
                 this.connectionManager.getClientUuid()
             );
 
-            this.invocationService.invokeOnConnection(connection, requestMessage).then(clientMessage => {
+            return this.invocationService.invokeOnConnection(connection, requestMessage).then(clientMessage => {
                 SqlServiceImpl.handleExecuteResponse(clientMessage, res);
+                return res;
             }).catch(err => {
-                res.onExecuteError(
-                    this.rethrow(err, connection)
-                );
+                const error = this.rethrow(err, connection);
```

Comment:
I think this logic is a bit flawed. Imagine that you executed a query and received an error (but like valid response with `response_error`, so you throw it from `handleExecuteResponse` above). With this catch, we will also catch it. Not only we will call `onExecuteError` (no problem, just not ideal) a second time, but if the connection is closed in between, we will report to user a `Cluster topology changed while a query was executed` error, while in fact it was a different error.

I believe you should do invokeOnConnection().catch().then()

Reply:
Now we call onExecuteError once like my java pr https://github.com/hazelcast/hazelcast/pull/19612/files